packages/compiler-core/src/transforms/transformExpression.ts

```diff
@@ -44,7 +44,9 @@ import { parseExpression } from '@babel/parser'
 import { IS_REF, UNREF } from '../runtimeHelpers'
 import { BindingTypes } from '../options'
 
-const isLiteralWhitelisted = /*@__PURE__*/ makeMap('true,false,null,this')
+const isLiteralWhitelisted = /*@__PURE__*/ makeMap(
+  'true,false,null,undefined,this',
```

Comment:
`undefined` can be used as a variable name, but we check if it's a variable everywhere we use `isLiteralWhitelisted`, so I believe this change is sound.